File: perceptron.ml
(* auxiliar absolute value *)
let absolute_float x =
    if (x < 0.0) then -1.0 *. x else x ;;

(* relu *)
let relu x = max(0,x);;

(* sigmoid *)
let sigmoid x = 1.0 /. (1.0 +. 2.71828 ** x) ;;

(* sigmoid derivative *)
let sigmoid_derivative x =
    sigmoid x *. (1.0 -. sigmoid x) ;;

(* auxiliar function to multiply elements of the tuples *)
let untuple (n, m) = n *. m ;;

(* neuron function *)
let neuron activation_function bias inputs =
    activation_function (bias +. List.fold_left (+.) 0.0 (List.map untuple inputs));;

(* delta rule *)
(*  n -> learning rate
    t -> target output
    y -> obtained output
    x -> input*)
let delta_rule n error x (weight, value) =
    (weight +. n *. (error) *. x, value) ;;

(* weight modification *)
let learn inputs error =
    List.map (delta_rule 0.1 error 0.1) inputs ;;

(* training function *)
let rec train inputs expected_result error_threshold =
    let error = expected_result -. (neuron sigmoid 1.0 inputs) in
    if (absolute_float error > error_threshold) 
        then train (learn inputs error) expected_result error_threshold
    else inputs
;;

File: perceptron.hs
-- auxiliar absolute value
absolute_float x =
    if (x < 0.0) then -1.0 * x else x

-- relu
relu x = max(0,x)

-- sigmoid
sigmoid x = 1.0/(1+exp(-x))

-- sigmoid derivative
sigmoid_derivative x =
    sigmoid x * (1.0 - sigmoid x )

-- auxiliar for inputs
untuple (weight, value) = weight * value

-- neuron function
neuron activation_function inputs bias =
    activation_function ( bias + (foldl (+) 0.0 (map untuple inputs)))

-- delta rule
delta_rule n error x (weight, value) =
    (weight + n * error * x, value)

-- weight modification
learn inputs error =
    map (delta_rule 0.1 error 0.1) inputs

-- training function
train inputs expected_result error_threshold =
    let error = expected_result - (neuron sigmoid inputs 1.0) in
    if (absolute_float error > error_threshold)
        then train (learn inputs error) expected_result error_threshold
    else inputs


-- main
main :: IO ()
main = return ()
